python/ray/experimental/ddp/ddp.py

```diff
@@ -0,0 +1,518 @@
+    mp.spawn(
+        torch_dist_run,
+        args=(world_size, torch_weights, config),
+        nprocs=world_size,
+        join=True,
+    )
+
+
+def torch_dist_run(rank, world_size, torch_weights, config: Config):
+    """Run DDP with `torch.distributed`."""
+    print(f"Running torch DDP on rank {rank}.")
+    os.environ["MASTER_ADDR"] = "localhost"
+    os.environ["MASTER_PORT"] = "12355"
+
+    # initialize the process group
+    dist.init_process_group("gloo", rank=rank, world_size=world_size)
+
+    # create model and move it to GPU with id rank
+    model = LayeredModel(
+        config.layer_size, config.num_layers, f"cuda:{rank}", config.dtype, config.lr
+    ).to(rank)
+    ddp_model = DDP(model, device_ids=[rank])
+
+    loss_fn = model.criterion
+    optimizer = optim.SGD(ddp_model.parameters(), lr=model.lr)
+
+    num_actors = config.num_actors
+
+    torch_ddp_weights = []
+    elapses = []
+    for i in range(config.it):
+        x, y = generate_input_output(config)
+        xs = torch.tensor_split(x, num_actors)
+        ys = torch.tensor_split(y, num_actors)
+        x = xs[rank]
+        y = ys[rank]
+        x = x.to(rank)
+        y = y.to(rank)
+        start = time.perf_counter()
+        optimizer.zero_grad()
+        outputs = ddp_model(x)
+        labels = y
+        loss = loss_fn(outputs, labels)
+        loss.backward()
+        optimizer.step()
+        end = time.perf_counter()
+
+        cur_iter_weights = []
+        for i in range(0, len(model.layers), 2):
+            layer: torch.nn.Linear = model.layers[i]
+            cur_iter_weights.append(torch.clone(layer.weight))
+        torch_ddp_weights.append(cur_iter_weights)
+
+        elapse = end - start
+        elapses.append(elapse)
+
+    print_elapses(elapses, f"torch ddp #{rank}")
+
+    dist.destroy_process_group()
+    compare_weights(torch_weights, torch_ddp_weights, "torch vs torch ddp")
+
+
+def main(config: Config) -> None:
+    ray.init()
+    if sum(node["Resources"].get("GPU", 0) for node in ray.nodes()) < config.num_actors:
+        print(f"Needs at least {config.num_actors} GPUs")
+        return
+
+    adag_ddp_weights = run_adag_ddp(config)
+
+    ray.shutdown()
+
+    check_correctness(adag_ddp_weights, config)
+
+
+def cli() -> Config:
+    str_to_dtype = {
+        "float32": torch.float32,
+        "float": torch.float,  # alias for float32
+        "float64": torch.float64,
+        "double": torch.double,  # alias for float64
+        "float16": torch.float16,
+        "half": torch.half,  # alias for float16
+    }
+    parser = argparse.ArgumentParser(
+        description="DDP demo (aDAG DDP vs torch vs torch DDP)"
+    )
+    parser.add_argument("--num-layers", type=int, help="number of layers")
+    parser.add_argument(
+        "--layer-size", type=int, help="size of a layer (each layer is a square)"
+    )
+    parser.add_argument(
+        "--dtype",
+        type=str,
+        choices=list(str_to_dtype.keys()),
```

Comment:
You already limit `choices`. I guess `args["dtype"] = str_to_dtype.get(args["dtype"], None)` is not needed.

Reply:
Changed to `str_to_dtype[args.dtype]`.